File: app/src/main/java/com/example/ui/ReaderViewScreen.kt
package com.example.ui

import android.graphics.Bitmap
import android.graphics.ImageDecoder
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Autorenew
import androidx.compose.material.icons.filled.BarChart
import androidx.compose.material.icons.filled.Bookmarks
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.CloudSync
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.DeleteSweep
import androidx.compose.material.icons.filled.Psychology
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Screenshot
import androidx.compose.material.icons.filled.Link
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Share
import androidx.compose.material.icons.filled.Timer
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.window.Dialog
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.combinedClickable
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Label
import androidx.compose.material.icons.filled.Menu
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.outlined.BarChart
import androidx.compose.material.icons.outlined.Bookmarks
import androidx.compose.material.icons.outlined.Settings
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.IconButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.ui.draw.clip
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.isImeVisible
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.expandVertically
import androidx.compose.animation.shrinkVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.material.icons.filled.ExpandLess
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.AutoAwesome
import android.content.Intent
import kotlinx.coroutines.launch
import com.example.MainActivity
import com.example.domain.model.AuthState
import com.example.domain.model.Bookmark
import com.example.domain.model.SourceType
import com.example.ui.components.GlassBottomBar
import com.example.ui.components.GlassNavigationItem
import com.example.ui.components.GlassScaffold
import com.example.ui.components.GlassTopBar
import com.example.ui.components.LiquidGlassFab
import com.example.ui.screens.auth.AuthViewModel
import com.example.ui.screens.auth.LoginScreen
import com.example.ui.theme.BookmarkTheme
import com.example.ui.theme.GlassTier
import com.example.ui.theme.glassSurface
import com.example.ui.theme.rememberGlassTier
import com.example.ui.theme.CurioMotion
import com.example.ui.theme.pressBounce
import com.example.ui.theme.bounceScale
import com.example.ui.theme.curioAccentBrush
import coil.compose.AsyncImage
import coil.compose.SubcomposeAsyncImage
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.offset
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.alpha
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.tween
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.icons.filled.MoreHoriz
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Article
import androidx.compose.material.icons.filled.TrendingUp
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.filled.Hub
import androidx.compose.material.icons.automirrored.filled.MenuBook
import java.text.SimpleDateFormat
import java.util.Locale

@OptIn(ExperimentalLayoutApi::class)
@Composable
fun ReaderViewScreen(
    bookmark: Bookmark,
    onClose: () -> Unit,
    tier: GlassTier,
    darkTheme: Boolean,
    dynamicColor: Boolean
) {
    var fontScale by remember { mutableStateOf(1.0f) } // 1.0 = Medium, 0.85 = Small, 1.25 = Large
    val context = LocalContext.current
    val formattedTime = SimpleDateFormat("MMM dd, yyyy - HH:mm", Locale.getDefault()).format(java.util.Date(bookmark.createdAt))

    androidx.compose.ui.window.Dialog(
        properties = androidx.compose.ui.window.DialogProperties(
            usePlatformDefaultWidth = false
        ),
        onDismissRequest = onClose
    ) {
        BookmarkTheme(darkTheme = darkTheme, dynamicColor = dynamicColor) {
            // Slide the reader up from the bottom edge to match the app's card modals.
            var visible by remember { mutableStateOf(false) }
            LaunchedEffect(Unit) { visible = true }
            AnimatedVisibility(
                visible = visible,
                enter = slideInVertically(
                    initialOffsetY = { it },
                    animationSpec = tween(durationMillis = 360)
                ) + fadeIn(animationSpec = tween(durationMillis = 240)),
                exit = fadeOut()
            ) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(MaterialTheme.colorScheme.background)
                    .safeDrawingPadding()
            ) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
            ) {
                // Header Top Bar
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 20.dp, vertical = 14.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    IconButton(
                        onClick = onClose,
                        modifier = Modifier.testTag("reader_back_button")
                    ) {
                        Icon(
                            imageVector = Icons.Default.ArrowBack,
                            contentDescription = "Go back to bookmarks",
                            tint = MaterialTheme.colorScheme.onSurface
                        )
                    }

                    Text(
                        text = "READER VIEW",
                        style = MaterialTheme.typography.labelMedium.copy(
                            fontWeight = FontWeight.ExtraBold,
                            color = MaterialTheme.colorScheme.primary,
                            letterSpacing = 1.5.sp
                        )
                    )

                    // Font Sizing adjusters
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(
                            text = "A-",
                            style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.Bold),
                            modifier = Modifier
                                .clickable { fontScale = 0.85f }
                                .padding(8.dp)
                                .testTag("reader_zoom_out")
                        )
                        Text(
                            text = "A",
                            style = MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.Medium),
                            modifier = Modifier
                                .clickable { fontScale = 1.0f }
                                .padding(8.dp)
                                .testTag("reader_zoom_reset")
                        )
                        Text(
                            text = "A+",
                            style = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.ExtraBold),
                            modifier = Modifier
                                .clickable { fontScale = 1.25f }
                                .padding(8.dp)
                                .testTag("reader_zoom_in")
                        )
                    }
                }

                // Main distraction-free scroll column constrained for tablet reading comfort
                Column(
                    modifier = Modifier
                        .align(Alignment.CenterHorizontally)
                        .widthIn(max = 600.dp)
                        .padding(horizontal = 24.dp, vertical = 16.dp),
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    // Category & Date metadata
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Box(
                            modifier = Modifier
                                .background(
                                    color = MaterialTheme.colorScheme.primary.copy(alpha = 0.12f),
                                    shape = RoundedCornerShape(8.dp)
                                )
                                .padding(horizontal = 10.dp, vertical = 4.dp)
                        ) {
                            Text(
                                text = (bookmark.category ?: "Uncategorized").uppercase(),
                                style = MaterialTheme.typography.labelSmall.copy(
                                    fontWeight = FontWeight.Black,
                                    color = MaterialTheme.colorScheme.primary
                                )
                            )
                        }

                        Text(
                            text = "•",
                            style = MaterialTheme.typography.titleSmall,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f)
                        )

                        Text(
                            text = formattedTime,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f)
                        )
                    }

                    // Title
                    Text(
                        text = bookmark.title?.ifBlank { "Untitled Curio" } ?: "Untitled Curio",
                        style = MaterialTheme.typography.headlineLarge.copy(
                            fontWeight = FontWeight.Black,
                            lineHeight = 38.sp
                        ),
                        color = MaterialTheme.colorScheme.onSurface,
                        modifier = Modifier.testTag("reader_title")
                    )

                    // URL or domain link
                    bookmark.url?.ifBlank { null }?.let { urlString ->
                        Text(
                            text = urlString,
                            style = MaterialTheme.typography.bodySmall.copy(
                                textDecoration = androidx.compose.ui.text.style.TextDecoration.Underline
                            ),
                            color = MaterialTheme.colorScheme.secondary,
                            modifier = Modifier.clickable {
                                try {
                                    val intent = Intent(Intent.ACTION_VIEW, Uri.parse(urlString))
                                    context.startActivity(intent)
                                } catch (e: Exception) {
                                    // Ignored
                                }
                            }
                        )
                    }

                    Spacer(modifier = Modifier.height(8.dp))

                    // Source paper / repo metadata (Phase 8)
                    if (bookmark.sourceType != null) {
                        val srcColor = when (bookmark.sourceType) {
                            SourceType.ARXIV -> Color(0xFFB71C1C)
                            SourceType.GITHUB -> Color(0xFF1B5E20)
                            SourceType.HUGGING_FACE -> Color(0xFFF57F17)
                            else -> MaterialTheme.colorScheme.secondary
                        }
                        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                            if (!bookmark.sourceTitle.isNullOrBlank()) {
                                Text(
                                    text = bookmark.sourceTitle,
                                    style = MaterialTheme.typography.headlineSmall.copy(
                                        fontWeight = FontWeight.Bold,
                                        lineHeight = 28.sp,
                                        fontSize = (18 * fontScale).sp
                                    ),
                                    color = srcColor
                                )
                            }
                            if (!bookmark.sourceAuthors.isNullOrBlank()) {
                                Text(
                                    text = bookmark.sourceAuthors,
                                    style = MaterialTheme.typography.bodySmall.copy(fontStyle = androidx.compose.ui.text.font.FontStyle.Italic),
                                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.65f)
                                )
                            }
                            if (bookmark.referenceCount > 1) {
                                Text(
                                    text = "Mentioned in ${bookmark.referenceCount} bookmarks",
                                    style = MaterialTheme.typography.labelSmall,
                                    color = srcColor.copy(alpha = 0.7f)
                                )
                            }
                            if (!bookmark.sourceAbstract.isNullOrBlank()) {
                                androidx.compose.material3.HorizontalDivider(color = srcColor.copy(alpha = 0.15f))
                                Text(
                                    text = "ABSTRACT",
                                    style = MaterialTheme.typography.labelSmall.copy(
                                        fontWeight = FontWeight.Black, color = srcColor, letterSpacing = 1.sp
                                    )
                                )
                                Text(
                                    text = bookmark.sourceAbstract,
                                    style = MaterialTheme.typography.bodyMedium.copy(
                                        fontSize = (14 * fontScale).sp,
                                        lineHeight = (22 * fontScale).sp,
                                        fontFamily = FontFamily.Serif
                                    ),
                                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.85f)
                                )
                            }
                        }
                        Spacer(modifier = Modifier.height(8.dp))
                    }

                    // Deep analysis (Phase 9)
                    if (bookmark.isDeepAnalyzed && !bookmark.deepSummary.isNullOrBlank()) {
                        androidx.compose.material3.HorizontalDivider(color = MaterialTheme.colorScheme.tertiary.copy(alpha = 0.2f))
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(
                            "DEEP ANALYSIS",
                            style = MaterialTheme.typography.labelSmall.copy(
                                fontWeight = FontWeight.Black,
                                color = MaterialTheme.colorScheme.tertiary,
                                letterSpacing = 1.sp
                            )
                        )
                        Text(
                            text = bookmark.deepSummary,
                            style = MaterialTheme.typography.bodyMedium.copy(
                                fontSize = (14 * fontScale).sp,
                                lineHeight = (22 * fontScale).sp
                            ),
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.85f)
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                    }

                    // Summarized contents block with optimized typography
                    Text(
                        text = bookmark.summary ?: "No Summary available. Execute AI analysis to generate a cognitive outline.",
                        style = MaterialTheme.typography.bodyLarge.copy(
                            fontSize = (16 * fontScale).sp,
                            lineHeight = (26 * fontScale).sp,
                            fontFamily = FontFamily.Serif
                        ),
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.85f),
                        modifier = Modifier
                            .fillMaxWidth()
                            .testTag("reader_summary_content")
                    )

                    // Tags list row
                    if (bookmark.tags.isNotEmpty()) {
                        Spacer(modifier = Modifier.height(16.dp))
                        Text(
                            "TAGS",
                            style = MaterialTheme.typography.labelSmall.copy(
                                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f),
                                fontWeight = FontWeight.Bold,
                                letterSpacing = 1.sp
                            )
                        )
                        FlowRow(
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                            verticalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            bookmark.tags.forEach { tag ->
                                Box(
                                    modifier = Modifier
                                        .background(
                                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.05f),
                                            shape = RoundedCornerShape(8.dp)
                                        )
                                        .padding(horizontal = 10.dp, vertical = 6.dp)
                                ) {
                                    Text(
                                        text = "#$tag",
                                        style = MaterialTheme.typography.bodySmall.copy(fontWeight = FontWeight.Bold),
                                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f)
                                    )
                                }
                            }
                        }
                    }

                    // OCR plain text content block (distraction free layout)
                    if (!bookmark.ocrText.isNullOrBlank()) {
                        Spacer(modifier = Modifier.height(24.dp))
                        androidx.compose.material3.HorizontalDivider(color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.1f))
                        Spacer(modifier = Modifier.height(16.dp))

                        Text(
                            "RAW TEXT TRANSCRIPTION",
                            style = MaterialTheme.typography.labelSmall.copy(
                                color = MaterialTheme.colorScheme.primary,
                                fontWeight = FontWeight.Black,
                                letterSpacing = 1.sp
                            )
                        )

                        Text(
                            text = bookmark.ocrText ?: "",
                            style = MaterialTheme.typography.bodyMedium.copy(
                                fontSize = (14 * fontScale).sp,
                                lineHeight = (22 * fontScale).sp
                            ),
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f),
                            modifier = Modifier
                                .fillMaxWidth()
                                .testTag("reader_raw_text")
                        )
                    }

                    Spacer(modifier = Modifier.height(80.dp))
                }
            }
            }
        }
    }
}
}

File: ios/Curio/Curio/Screens/ReaderViewScreen.swift
//
//  ReaderViewScreen.swift
//  Curio
//
//  Ports: app/src/main/java/com/example/ui/ReaderViewScreen.kt (ReaderViewScreen).
//
//  DESIGN §10 (Screens): Full-screen reader (font scale, metadata, source/deep/summary/tags/OCR
//  blocks, 600pt cap). `struct ReaderView`; `.fullScreenCover`; slide-up transition; `@State
//  fontScale` (0.85/1.0/1.25); serif body; source-type colors; date `MMM dd, yyyy - HH:mm`; openURL.
//
//  CONVENTIONS mapping:
//  - §8 "Theme": the Android `BookmarkTheme(darkTheme, dynamicColor)` wrapper is replaced by
//    `.curioTheme(darkTheme:)` (Material You / `dynamicColor` is DROPPED — the static Cosmic palette
//    is canonical), so the reader forces its own theme so the cover matches the app's resolved scheme
//    even though `.fullScreenCover` presents in a fresh environment. The source-type accent colors
//    (`Color(0xFFB71C1C)` arXiv etc.) are unpacked at this UI boundary.
//  - §8 "Layout/insets": the Android `Dialog(usePlatformDefaultWidth = false)` full-bleed reader maps
//    to `.fullScreenCover` (hosted by `BookmarkApp`). This file is the cover CONTENT; the slide-up
//    + fade entrance transition mirrors the Compose `AnimatedVisibility(slideInVertically + fadeIn)`.
//    `safeDrawingPadding()` → `.safeAreaPadding()` semantics via the cover's safe area.
//  - §4 "Sealed strings": exact user-facing strings preserved verbatim — "READER VIEW",
//    "Uncategorized", "Untitled Curio", "No Summary available. Execute AI analysis to generate a
//    cognitive outline.", "ABSTRACT", "DEEP ANALYSIS", "TAGS", "RAW TEXT TRANSCRIPTION",
//    "Mentioned in {n} bookmarks".
//  - UI INVARIANT (CONVENTIONS / Bookmark.swift): the Android reader DOES surface `category` in the
//    metadata chip (it predates the "category never in UI" rule for the feed). This faithful port
//    keeps that one display verbatim — `(bookmark.category ?? "Uncategorized").uppercased()` —
//    matching the Kotlin source byte-for-byte (the no-category-in-UI rule governs the FEED/cards).
//  - §8 "Motion" / Accessibility: the back/zoom controls are 48pt targets with `.accessibilityLabel`
//    + `.accessibilityIdentifier` carried over from the Compose `semantics { }` / `testTag(...)`.
//

import SwiftUI

/// Full-screen, distraction-free reader for a single bookmark. Direct port of `@Composable
/// ReaderViewScreen`. Presented by `BookmarkApp` via `.fullScreenCover(item:)`; `onClose` tears the
/// cover down. `darkTheme` forces the reader's own theme wrapper so the cover matches the resolved
/// app scheme.
struct ReaderView: View {
    let bookmark: Bookmark
    let tier: GlassTier
    let darkTheme: Bool
    let onClose: () -> Void

    private static let fontScaleKey = "reader_font_scale"

    /// `1.0 = Medium, 0.85 = Small, 1.25 = Large` — persisted like Android `curio_reader` prefs.
    @State private var fontScale: CGFloat = {
        let stored = UserDefaults.standard.float(forKey: fontScaleKey)
        return stored > 0 ? CGFloat(stored) : 1.0
    }()
    /// Drives the slide-up + fade entrance (Compose `AnimatedVisibility` `visible` flag).
    @State private var visible: Bool = false

    @Environment(\.openURL) private var openURL

    /// `SimpleDateFormat("MMM dd, yyyy - HH:mm", Locale.getDefault()).format(Date(createdAt))`.
    /// `createdAt` is epoch milliseconds (CONVENTIONS §6); the formatter uses the current locale and
    /// device time zone, matching `Locale.getDefault()` + the default `SimpleDateFormat` zone.
    private var formattedTime: String {
        let date = Date(timeIntervalSince1970: Double(bookmark.createdAt) / 1000.0)
        let formatter = DateFormatter()
        formatter.locale = Locale.current
        formatter.dateFormat = "MMM dd, yyyy - HH:mm"
        return formatter.string(from: date)
    }

    var body: some View {
        ReaderColors { colors in
            ZStack {
                colors.background
                    .ignoresSafeArea()

                if visible {
                    content(colors: colors)
                        .transition(
                            .move(edge: .bottom).combined(with: .opacity)
                        )
                }
            }
            .background(colors.background)
        }
        // Force the reader's own resolved scheme so the cover (presented in a fresh environment)
        // matches the app (Compose `BookmarkTheme(darkTheme, dynamicColor)`).
        .curioTheme(darkTheme: darkTheme)
        .onAppear {
            // `LaunchedEffect(Unit) { visible = true }` — slide the reader up on first composition.
            withAnimation(.easeOut(duration: 0.36)) { visible = true }
        }
        .onChange(of: fontScale) { _, scale in
            UserDefaults.standard.set(Float(scale), forKey: Self.fontScaleKey)
        }
    }

    // MARK: - Content

    @ViewBuilder
    private func content(colors: CurioColorScheme) -> some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                headerBar(colors: colors)

                // Main distraction-free scroll column constrained for tablet reading comfort.
                VStack(alignment: .leading, spacing: 16) {
                    metadataRow(colors: colors)
                    titleBlock(colors: colors)
                    urlBlock(colors: colors)

                    Spacer().frame(height: 8)

                    sourceBlock(colors: colors)
                    deepAnalysisBlock(colors: colors)
                    summaryBlock(colors: colors)
                    tagsBlock(colors: colors)
                    ocrBlock(colors: colors)

                    Spacer().frame(height: 80)
                }
                .padding(.horizontal, 24)
                .padding(.vertical, 16)
                .frame(maxWidth: 600)
                .frame(maxWidth: .infinity, alignment: .center)
            }
        }
    }

    // MARK: - Header bar (back + READER VIEW + font sizing)

    @ViewBuilder
    private func headerBar(colors: CurioColorScheme) -> some View {
        HStack {
            Button(action: onClose) {
                Image(systemName: "arrow.left")
                    .font(.system(size: 18))
                    .foregroundStyle(colors.onSurface)
                    .frame(width: 44, height: 44)
                    .contentShape(Rectangle())
                    .accessibilityLabel("Go back to bookmarks")
            }
            .buttonStyle(.curioPressBounce)
            .accessibilityIdentifier("reader_back_button")

            Spacer()

            Text("READER VIEW")
                .font(.system(size: 14, weight: .heavy))
                .tracking(1.5)
                .foregroundStyle(colors.primary)

            Spacer()

            // Font-sizing adjusters (A- / A / A+), each a 48pt button.
            HStack(spacing: 4) {
                fontButton(
                    label: "A-",
                    size: 14, weight: .bold,
                    accessibility: "Decrease font size",
                    identifier: "reader_zoom_out",
                    colors: colors,
                    selected: abs(fontScale - 0.85) < 0.01
                ) { fontScale = 0.85 }

                fontButton(
                    label: "A",
                    size: 16, weight: .medium,
                    accessibility: "Reset font size",
                    identifier: "reader_zoom_reset",
                    colors: colors,
                    selected: abs(fontScale - 1.0) < 0.01
                ) { fontScale = 1.0 }

                fontButton(
                    label: "A+",
                    size: 22, weight: .heavy,
                    accessibility: "Increase font size",
                    identifier: "reader_zoom_in",
                    colors: colors,
                    selected: abs(fontScale - 1.25) < 0.01
                ) { fontScale = 1.25 }
            }
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 14)
    }

    @ViewBuilder
    private func fontButton(
        label: String, size: CGFloat, weight: Font.Weight,
        accessibility: String, identifier: String,
        colors: CurioColorScheme,
        selected: Bool = false,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            Text(label)
                .font(.system(size: size, weight: weight))
                .foregroundStyle(selected ? colors.primary : colors.onSurface)
                .padding(8)
                .frame(minWidth: 48, minHeight: 48)
                .background(
                    selected ? colors.primary.opacity(0.14) : Color.clear,
                    in: RoundedRectangle(cornerRadius: 8, style: .continuous)
                )
                .contentShape(Rectangle())
                .accessibilityLabel(accessibility)
        }
        .buttonStyle(.curioPressBounce)
        .accessibilityIdentifier(identifier)
    }

    // MARK: - Metadata row (category chip · date)

    @ViewBuilder
    private func metadataRow(colors: CurioColorScheme) -> some View {
        HStack(spacing: 8) {
            // Faithful to the Android reader: the AI category IS shown here in the reader metadata
            // chip (the "no category in UI" rule governs the feed cards).
            Text((bookmark.category ?? "Uncategorized").uppercased())
                .font(.system(size: 11, weight: .black))
                .foregroundStyle(colors.primary)
                .padding(.horizontal, 10)
                .padding(.vertical, 4)
                .background(colors.primary.opacity(0.12), in: RoundedRectangle(cornerRadius: 8, style: .continuous))

            Text("•")
                .font(.system(size: 14))
                .foregroundStyle(colors.onSurface.opacity(0.4))

            Text(formattedTime)
                .font(.system(size: 12, weight: .regular))
                .foregroundStyle(colors.onSurface.opacity(0.5))
        }
    }

    // MARK: - Title

    @ViewBuilder
    private func titleBlock(colors: CurioColorScheme) -> some View {
        // `bookmark.title?.ifBlank { "Untitled Curio" } ?: "Untitled Curio"`.
        let title: String = {
            if let t = bookmark.title, !t.isBlank { return t }
            return "Untitled Curio"
        }()
        Text(title)
            .font(.system(size: 32, weight: .black))
            .lineSpacing(38 - 32)
            .foregroundStyle(colors.onSurface)
            .frame(maxWidth: .infinity, alignment: .leading)
            .accessibilityIdentifier("reader_title")
    }

    // MARK: - URL / domain link

    @ViewBuilder
    private func urlBlock(colors: CurioColorScheme) -> some View {
        // `bookmark.url?.ifBlank { null }?.let { … }`.
        if let raw = bookmark.url, !raw.isBlank {
            Text(raw)
                .font(.system(size: 12, weight: .regular))
                .underline()
                .foregroundStyle(colors.secondary)
                .onTapGesture {
                    // try { startActivity(ACTION_VIEW, url) } catch { /* ignored */ }
                    if let url = URL(string: raw) {
                        openURL(url)
                    }
                }
        }
    }

    // MARK: - Source paper / repo metadata (Phase 8)

    @ViewBuilder
    private func sourceBlock(colors: CurioColorScheme) -> some View {
        if let sourceType = bookmark.sourceType {
            // Source-type accent — unpacked at the UI boundary.
            let srcColor: Color = {
                switch sourceType {
                case .ARXIV: return Color(argb: 0xFFB71C1C)
                case .GITHUB: return Color(argb: 0xFF1B5E20)
                case .HUGGING_FACE: return Color(argb: 0xFFF57F17)
                default: return colors.secondary
                }
            }()

            VStack(alignment: .leading, spacing: 8) {
                if let sourceTitle = bookmark.sourceTitle, !sourceTitle.isBlank {
                    Text(sourceTitle)
                        .font(.system(size: 18 * fontScale, weight: .bold))
                        .lineSpacing(28 - 18)
                        .foregroundStyle(srcColor)
                }
                if let sourceAuthors = bookmark.sourceAuthors, !sourceAuthors.isBlank {
                    Text(sourceAuthors)
                        .font(.system(size: 12, weight: .regular))
                        .italic()
                        .foregroundStyle(colors.onSurface.opacity(0.65))
                }
                if bookmark.referenceCount > 1 {
                    Text("Mentioned in \(bookmark.referenceCount) bookmarks")
                        .font(.system(size: 11, weight: .regular))
                        .foregroundStyle(srcColor.opacity(0.7))
                }
                if let sourceAbstract = bookmark.sourceAbstract, !sourceAbstract.isBlank {
                    Divider().overlay(srcColor.opacity(0.15))
                    Text("ABSTRACT")
                        .font(.system(size: 11, weight: .black))
                        .tracking(1.0)
                        .foregroundStyle(srcColor)
                    Text(sourceAbstract)
                        .font(.system(size: 14 * fontScale, weight: .regular, design: .serif))
                        .lineSpacing((22 - 14) * fontScale)
                        .foregroundStyle(colors.onSurface.opacity(0.85))
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            Spacer().frame(height: 8)
        }
    }

    // MARK: - Deep analysis (Phase 9)

    @ViewBuilder
    private func deepAnalysisBlock(colors: CurioColorScheme) -> some View {
        if bookmark.isDeepAnalyzed, let deep = bookmark.deepSummary, !deep.isBlank {
            Divider().overlay(colors.tertiary.opacity(0.2))
            Spacer().frame(height: 8)
            Text("DEEP ANALYSIS")
                .font(.system(size: 11, weight: .black))
                .tracking(1.0)
                .foregroundStyle(colors.tertiary)
            Text(deep)
                .font(.system(size: 14 * fontScale, weight: .regular))
                .lineSpacing((22 - 14) * fontScale)
                .foregroundStyle(colors.onSurface.opacity(0.85))
            Spacer().frame(height: 8)
        }
    }

    // MARK: - Summary

    @ViewBuilder
    private func summaryBlock(colors: CurioColorScheme) -> some View {
        // `bookmark.summary ?: "No Summary available. …"`.
        let summary = bookmark.summary ?? "No Summary available. Execute AI analysis to generate a cognitive outline."
        Text(summary)
            .font(.system(size: 16 * fontScale, weight: .regular, design: .serif))
            .lineSpacing((26 - 16) * fontScale)
            .foregroundStyle(colors.onSurface.opacity(0.85))
            .frame(maxWidth: .infinity, alignment: .leading)
            .accessibilityIdentifier("reader_summary_content")
    }

    // MARK: - Tags

    @ViewBuilder
    private func tagsBlock(colors: CurioColorScheme) -> some View {
        if !bookmark.tags.isEmpty {
            Spacer().frame(height: 16)
            Text("TAGS")
                .font(.system(size: 11, weight: .bold))
                .tracking(1.0)
                .foregroundStyle(colors.onSurface.opacity(0.4))
            FlowLayout(spacing: 8) {
                ForEach(bookmark.tags, id: \.self) { tag in
                    Text("#\(tag)")
                        .font(.system(size: 12, weight: .bold))
                        .foregroundStyle(colors.onSurface.opacity(0.7))
                        .padding(.horizontal, 10)
                        .padding(.vertical, 6)
                        .background(colors.onSurface.opacity(0.05), in: RoundedRectangle(cornerRadius: 8, style: .continuous))
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }

    // MARK: - OCR raw text (distraction-free)

    @ViewBuilder
    private func ocrBlock(colors: CurioColorScheme) -> some View {
        if let ocr = bookmark.ocrText, !ocr.isBlank {
            Spacer().frame(height: 24)
            Divider().overlay(colors.onSurface.opacity(0.1))
            Spacer().frame(height: 16)

            Text("RAW TEXT TRANSCRIPTION")
                .font(.system(size: 11, weight: .black))
                .tracking(1.0)
                .foregroundStyle(colors.primary)

            Text(ocr)
                .font(.system(size: 14 * fontScale, weight: .regular))
                .lineSpacing((22 - 14) * fontScale)
                .foregroundStyle(colors.onSurface.opacity(0.7))
                .frame(maxWidth: .infinity, alignment: .leading)
                .accessibilityIdentifier("reader_raw_text")
        }
    }
}

// MARK: - ReaderColors

/// Reads the active `CurioColorScheme` (resolved by the reader's own `.curioTheme(darkTheme:)`
/// wrapper) and hands it to a builder — used so the body can pass `colors` into the block helpers
/// without each one independently reading the environment. Equivalent to inlining
/// `@Environment(\.curioColors)`.
private struct ReaderColors<Content: View>: View {
    @Environment(\.curioColors) private var colors
    @ViewBuilder let content: (CurioColorScheme) -> Content
    var body: some View { content(colors) }
}

// Note: the wrapping tag layout reuses the shared `FlowLayout` (Compose `FlowRow` analogue)
// declared in `Components/CurioPostCard.swift`.
